UM/OutputDevice/OutputDeviceManager.py

```diff
@@ -218,8 +265,58 @@ def removeOutputDevicePlugin(self, plugin_id: str) -> None:
     def getOutputDevicePlugin(self, plugin_id: str) -> Optional["OutputDevicePlugin"]:
         return self._plugins.get(plugin_id, None)
 
+    ##  Sometimes automatic discovery doesn't or can't work, and a device has to be added 'manually' by some address.
+    #   This will loop through all plugins to find one that accepts the address.
+    #   Note that a plugin can claim priority for certain addresses, but this should be used with great care!
+    #   \param address The address of this device, often an IP or similar.
+    #   \param plugin_types Limit the search to these plugins (or empty list for 'accept all', which is the default).
+    def addManualDevice(self, address: str, plugin_types: Optional[List[str]] = None) -> None:
+        priority_order = [
+                ManualDeviceAdditionAttempt.PRIORITY,
+                ManualDeviceAdditionAttempt.POSSIBLE
+            ]  # type: List[ManualDeviceAdditionAttempt]
+        plugins_by_priority = {}  # type: Dict[ManualDeviceAdditionAttempt, List[str]]
+
+        for plugin_id, plugin in self._plugins.items():
+            if plugin and (not plugin_types or plugin_id in plugin_types):
+                can_add = plugin.canAddManualDevice(address)
+                if can_add not in plugins_by_priority:
+                    plugins_by_priority[can_add] = []
+                plugins_by_priority[can_add].append(plugin_id)
+
+        for priority in priority_order:
+            if priority not in plugins_by_priority:  # None of the plugins that specified 'NO' will be requested.
+                continue
+            for plugin_id in plugins_by_priority[priority]:
+                plugin = self._plugins[plugin_id]
+                if plugin:
+                    plugin.addManualDeviceSignal.connect(self._onManualDeviceAdded)
+                    plugin.removeManualDevice("", address)
+                    plugin.addManualDevice(address)
+                    return
```

Comment:
You could simplify this a bit with Python's built-ins for sorting and filtering. It's probably faster, too. I'm thinking something like:

```
#Only allow plugins that are in the priority order.
accepted_plugins = list(filter(lambda plugin: plugin.canAddManualDevice(address) in priority_order, self._plugins.values()))
if not accepted_plugins:
    Logger.log("d", "Could not find a plugin blablabla")

return max(accepted_plugins, key = lambda plugin: priority_order.index(plugin.canAddManualDevice(address)))
```
(Not tested yet though.)

Reply:
Used your suggestion to halve the linecount of that part :-)

Not entirely happy about the two calls to `canAddManualDevice` instead of one, but overall it's better (and that function is explicitly there to _quickly_ check if that plugin can add the machine, so its fine).